algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -153,25 +308,28 @@ fn coordinate_descent<'a, F: Float>(
     while n_steps < max_steps {
         let mut w_max = F::zero();
         let mut d_w_max = F::zero();
-        for ii in 0..n_features {
-            if abs_diff_eq!(norm_cols_x[ii], F::zero()) {
+        for j in 0..n_features {
+            if abs_diff_eq!(norm_cols_x[j], F::zero()) {
                 continue;
             }
-            let w_ii = w[ii];
-            let x_slc: ArrayView1<F> = x.slice(s![.., ii]);
-            if abs_diff_ne!(w_ii, F::zero()) {
-                // FIXME: direct addition with loop might be faster as it does not have to allocate
-                r += &(&x_slc * w_ii);
+            let old_w_j = w[j];
+            let x_j: ArrayView1<F> = x.slice(s![.., j]);
+            if abs_diff_ne!(old_w_j, F::zero()) {
+                for i in 0..x.shape()[0] {
+                    r[i] += x_j[i] * old_w_j;
+                }
```

Comment:
Instead of adding in a loop you can do `r.scaled_add(w_ii, &x_slc)`. You can probably apply the same change in `block_coordinate_descent` as well.